runtime/descriptor.go

```diff
@@ -98,24 +105,66 @@ func propertySet(f *Frame, desc, inst, value *Object) *BaseException {
 
 // makeStructFieldDescriptor creates a descriptor with a getter that returns
 // the field given by fieldName from t's basis structure.
-func makeStructFieldDescriptor(t *Type, fieldName, propertyName string) *Object {
+func makeStructFieldDescriptor(t *Type, fieldName, propertyName string, fieldType fieldDescriptorType) *Object {
 	field, ok := t.basis.FieldByName(fieldName)
 	if !ok {
 		logFatal(fmt.Sprintf("no such field %q for basis %s", fieldName, nativeTypeName(t.basis)))
 	}
+
 	getterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
-		var ret *Object
-		var raised *BaseException
-		if raised = checkFunctionArgs(f, fieldName, args, ObjectType); raised == nil {
-			o := args[0]
-			if !o.isInstance(t) {
+		if raised := checkFunctionArgs(f, fieldName, args, ObjectType); raised != nil {
+			return nil, raised
+		}
+
+		self := args[0]
+		if !self.isInstance(t) {
+			format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
+			return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), self.typ.Name()))
+		}
+
+		return WrapNative(f, t.slots.Basis.Fn(self).FieldByIndex(field.Index))
+	}
+
+	setter := None
+	if fieldType == fieldDescriptorRW {
+		if field.PkgPath != "" {
+			logFatal(fmt.Sprintf("field '%q' is not public on Golang code. Please fix it.", fieldName))
+		}
+
+		setterFunc := func(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+			if raised := checkFunctionArgs(f, fieldName, args, ObjectType, ObjectType); raised != nil {
+				return nil, raised
+			}
+
+			self := args[0]
+			newValue := args[1]
+
+			if !self.isInstance(t) {
 				format := "descriptor '%s' for '%s' objects doesn't apply to '%s' objects"
-				raised = f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), o.typ.Name()))
-			} else {
-				ret, raised = WrapNative(f, t.slots.Basis.Fn(o).FieldByIndex(field.Index))
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, propertyName, t.Name(), self.typ.Name()))
+			}
+
+			if !newValue.isInstance(getNativeType(field.Type)) {
```

Comment:
Why is this necessary? This check should be handled implicitly by the maybeConvertValue() call below. As written, I believe that this will reject assignments between compatible types, like assigning to a field of `type Foo int` with an `int` value.